---
title: "Simulate SNP pattern"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simulate SNP pattern}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, setup-hidden}
#| include: false
knitr::opts_chunk$set(fig.alt = " ")
```

```{r, setup}
library(conflicted)
library(tekkamaki)
library(ggplot2)
grDevices::palette("Okabe-Ito")
options(
  ggplot2.continuous.colour = "viridis",
  ggplot2.continuous.fill = "viridis",
  ggplot2.discrete.colour = grDevices::palette()[-1],
  ggplot2.discrete.fill = grDevices::palette()[-1]
)
```

## Generate a gene genealogy

```{r, run}
set.seed(666L)
result = tekka("-y40 -K1000 -l2 --sa 2,2 --sj 2,2")
samples = result$sample_family[[1L]]
genealogy = make_gene_genealogy(samples) |> print()
```

## Visualize the generated gene genealogy

```{r, plot}
#| fig-height: 8
#| fig-width: 10
plot(genealogy) +
  theme_void() +
  theme(legend.position = "top")
```

## Generate SNPs

Randomly place a fixed number of mutations on a given genealogy without recombination.

```{r, snp}
snp = place_mutations(genealogy, 3L) |> print()
```

Save SNP patterns as a VCF file:

```{r, vcf}
#| echo: -1
some_file = stdout()
write_vcf(snp, some_file)
```

You can also generate a list of SNP data.frame for chromosomes with recombination in parallel.
Use `RNGkind("L'Ecuyer-CMRG")` and `set.seed()` for reproducibility:

```{r, snp-shortcut}
options(mc.cores = 2L) # parallel::detectCores(logical = FALSE)
RNGkind("L'Ecuyer-CMRG")
set.seed(666L)
snp = make_snp(samples, ss = c(chr7 = 4L, chr8 = 3L))
vcf = as_vcf(snp) |> add_pos_id()
write_vcf(vcf, some_file)
```
